### DISCUSSION TITLE
[Feature] Preact Support

### DISCUSSION THREAD

**Discussion by developit** [remix-run/remix#4562] [Proposals]
## Preact Support

- [x] Skip `<Scripts>` diffing on the client _(change in Remix)_
    - ~#390~
    - ~#403~
    - #422
- [x] Backport `<!DOCTYPE>` hydration skipping to Preact 10 _(change in Preact)_
    - preactjs/preact#3327
- [ ] Land patch for `document.documentElement` replacement _(change in `preact/compat`)_
    - preactjs/preact#3328

**Comment by JoviDeCroock** [comment]
preactjs/preact#3327 has been published in 10.6.0, we can also merge in the `document.documentElement` patch later unless we want to fix this another way?

**Comment by pspeter3** [comment]
I'm very excited about Preact support for Remix. What can I do to help?

**Comment by JoviDeCroock** [comment]
Hey @pspeter3 it should work as of now, the only thing we need to do different is the `hydrate` call. Rather than `hydrate(x, document)` we'll need to do

```js
const documentElement = document.documentElement;
const apply = n => document.replaceChild(n, documentElement);
hydrate(x,  {
	childNodes: [documentElement],
	firstChild: documentElement,
	insertBefore: apply,
	appendChild: apply
}
```

This because of our algo getting a bit confused 😅 we're hoping to land a fix but looking into the best way currently

https://twitter.com/_developit/status/1471294946560851969

**Comment by pspeter3** [comment]
Thanks! Is https://github.com/preactjs/preact/pull/3328 the right issue to follow from the Preact side?

**Comment by JoviDeCroock** [comment]
Well that was an initial solution but we're trying to find a smaller (and maybe in core) one

**Comment by arjunyel** [comment]
Is there an example repo with this work around used? Preact is awesome :D

> Hey @pspeter3 it should work as of now, the only thing we need to do different is the `hydrate` call. Rather than `hydrate(x, document)` we'll need to do
> 
> ```js
> const documentElement = document.documentElement;
> const apply = n => document.replaceChild(n, documentElement);
> hydrate(x,  {
> 	childNodes: [documentElement],
> 	firstChild: documentElement,
> 	insertBefore: apply,
> 	appendChild: apply
> }
> ```
> 
> This because of our algo getting a bit confused 😅 we're hoping to land a fix but looking into the best way currently
> 
> https://twitter.com/_developit/status/1471294946560851969

**Comment by phiberber** [comment]
After some hours of implementing Preact in Remix, I finally found how annoying not having access to the Bundler API can be. And even after patching Remix compiler.js file, it still didn't work because of compatibility between Preact 8 and Preact X, and I didn't manage to make Preact X work with ESBuild. It would be cool if Jovi guided us in the correct direction on implementing it in Remix, as Preact is a key factor in performance optimization.

**Comment by JoviDeCroock** [comment]
I would love to help out but we're not going to work on supporting 8 😅 else we'd be working on three major versions at once, I'll try to contribute a working example in the coming week

**Comment by JoviDeCroock** [comment]
@phiberber https://github.com/remix-run/remix/pull/1636 adds an example

**Comment by isaac-mcfadyen** [comment]
Is there any update on this? The example PR was closed with the status "we are waiting to get this natively supported".

**Comment by JoviDeCroock** [comment]
Haven't heard back about this 😅

**Comment by phiberber** [comment]
Jovi, should've waited for the official implementation before deleting the example branch, anyone can follow the commit steps though. 

It has been a couple months already, I'm not even sure if this is going to be released this year, and based on this issue's activity, seems like thete is little to none demand (compared to other issues).

**Comment by kiliman** [comment]
I do know the Remix team really wants to make Remix front-end agnostic. Preact has always been at the top of their list. With summer vacations and the major update to React Router, I believe this work was put on hold. It should pick back up soon after the new "Remix" router lands (with the data API) and deferred support.

**Comment by Klukies** [comment]
There's also [the remix routers](https://github.com/brophdawg11/remix-routers) repository, where work is being done to port `react-router-dom` based on the underlying `@remix-run/router`. I think it's currently on for Vue & Svelte is in the works.
I personally didn't commit to doing any work because of the statement made by Ryan about native support 😅.

There was also a [tweet](https://twitter.com/ryanflorence/status/1539240346839748608?s=20&t=T9adC_KjTFiGyMoMiahrwA) from Ryan himself playing around with Preact & HTM. He's importing what is a pre-release of the remix-router package, so I would presume there's some work that was already done.

**Comment by jacob-ebey** [comment]
@LukasCornille I've done an initial pass of a Preact layer here: https://github.com/brophdawg11/remix-routers/pull/19

**Comment by arjunyel** [comment]
Example with Deno https://twitter.com/ebey_jacob/status/1567055958727340032

**Comment by pspeter3** [comment]
What would it take to get official support at this point?

**Comment by developit** [comment]
[Streaming SSR support](https://github.com/preactjs/preact-render-to-string/pull/259).

**Comment by brophdawg11** [comment]
<img width="375" alt="Screenshot 2022-12-19 at 12 49 06 PM" src="https://user-images.githubusercontent.com/1609022/208488341-fa342356-0345-4013-900c-ed6866bc52b8.png">

😂 So many emoji reactions yet only 7 upvotes.  Remember the upvote (↑) is all that counts towards raising visibility on the roadmap so don't forget to upvote it if this is something you'd like to see worked on!

**Reply by nickluger** [reply]
Right, but that's in fact a UX problem caused by GitHub. 😕 People are naturally inclined to clicking on colorful icons (emojis) than on a grayish button.

**Comment by MichaelDeBoey** [comment]
As @ryanflorence pointed out in #7823, Remix is dedicated to be a React framework and not support other renderers

**Comment by mbrowne** [comment]
@developit Is there anything actually blocking official support for Preact at this point? It looks streaming SSR was added to Preact. I get that Preact support isn't a priority for the core Remix team at the moment, but if someone were to raise a PR, does anyone have a sense of if it would be accepted?

**Reply by jfsiii** [reply]
>  if someone were to raise a PR, does anyone have a sense of if it would be accepted?

Based on this https://github.com/remix-run/remix/discussions/7823 _**emphasis mine**_

> In the beginning of Remix, we also wanted to leave the door open for making the rendering library pluggable, hence @remix-run/react.
>
> Fast forward three years again and we've seen nearly every rendering lib implement their own remix-like framework with loaders, actions, adapters, nested routes, etc. We were interested in collaborating as a shared framework, but not interested in competing directly (we wish them well, they've done great work).
>
> We also really like the core ideas behind RSC in React. _**Making the rendering lib pluggable would prove to be a huge challenge and make Remix with React vs. something else very different. This is no longer a goal of the project**_

And signals I can't think of/find at the moment I get the strong sense it would not be accepted

**Reply by mbrowne** [reply]
Hmm, but Preact has a compatibility layer with react (`@preact/compat`), so it seems strange that the maintainers would be against it based on that argument alone. Hopefully one of the maintainers will weigh in.
